IMN-392: refactor logic agreement in agreement-process

AsterITA opened the pull request:
[To merge after #357 ]

his pull request is for refactoring the logic functions in `agreement-process`, unifying them with the main logic called from the endpoint and utilizing standards already adopted in other services

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -192,66 +277,162 @@ export function agreementServiceBuilder(
+      const [agreementdocumentSeed, updatesEvents] =
+        await processSubmitAgreement({
+          agreementData: agreement,
+          eservice: await retrieveEService(
+            agreement.data.eserviceId,
+            eserviceQuery
+          ),
```
I would move this above and assign it to a const instead of doing it inline here, to make the sequence of validations more explicit.

AsterITA replied:
> I would move this above and assign it to a const instead of doing it inline here, to make the sequence of validations more explicit.

I tried to avoid declaring const when they are only needed once but I agree it would be more clear 👍

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -192,66 +277,162 @@ export function agreementServiceBuilder(
+          payload,
+          agreementQuery,
+          tenantQuery,
+          constractBuilder: contractBuilder(
```
Typo, plus I would suggest the same here as well: instance it in a dedicated const above then pass it as param 

```suggestion
          contractBuilder: contractBuilder(
```

AsterITA replied:
I prefer to not create a const for this since `contractBuilder` is already a const

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -192,66 +277,162 @@ export function agreementServiceBuilder(
       }
 
-      return agreement;
+      return agreementdocumentSeed;
```
This function returns an `Agreement`, so I would rename this variable into `agreement` when you assign it above

AsterITA replied:
> This function returns an `Agreement`, so I would rename this variable into `agreement` when you assign it above

You're right, I renamed it in `submittedAgreement` since we already have a `agreement` in scope

ecamellini commented on packages/agreement-process/src/services/agreementSubmissionProcessor.ts:
```diff
@@ -18,107 +17,63 @@ import {
-  if (!eservice) {
-    throw eServiceNotFound(agreement.data.eserviceId);
-  }
+  const consumer = await retrieveTenant(agreement.consumerId, tenantQuery);
```
What's the reason to retrieve from readmodel here in the processor rather than in `agreementService.ts` file? I see that you retrieve Agreement and ESeervice in `agreementService.ts` before calling this function, while you retrieve Tenants (consumer & producer) here.

AsterITA replied:
> What's the reason to retrieve from readmodel here in the processor rather than in `agreementService.ts` file? I see that you retrieve Agreement and ESeervice in `agreementService.ts` before calling this function, while you retrieve Tenants (consumer & producer) here.

Here I made a mistake, `retrieveTenant` should be called after `validateSubmitOnDescriptor` because there's no need to retrieve the tenant if that function throws.

`Agreement` is needed in `agreementService.ts` for various validations before calling the processor while `Eservice` is used as soon as we enter the processor and we could retrieve it in it too but in order to do so I would need to pass to it the `EserviceQuery` parameter, so, in that case, I preferred to retrieve it just before calling the processor.

What do you think about this?

ecamellini replied:
If I understood correctly you retrieve tenants later in the code (in the processor file and not in the main `agreementService.ts`) so that the order of validations stays the same as before. If that's the case, it works for me - even though for consistency I would love to have all readmodel retrievals in the main service file where most of them are already :)

galales replied:
I had the same curiosity. 
I agree on the preference of delaying retrievals until they are actually needed, but `verifyConsumerDoesNotActivatePending` and `assertActivableState` are business logic checks and are in the service, while `validateActivationOnDescriptor` (and tenant retrieve) are in the process function.
In this case I don't see a clear pattern.

No strong opinions, I just would have expected them all in one function or the other

AsterITA replied:
I reorganized the code based on your feedbacks, let me know what you think @ecamellini @galales

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -324,17 +577,56 @@ export function agreementServiceBuilder(
-      await repository.createEvent(events);
 
-      return agreement;
+      const isProducer = authData.organizationId === agreement.data.producerId;
+      const isConsumer = authData.organizationId === agreement.data.consumerId;
+
+      if (!isProducer && !isConsumer) {
+        throw new Error(
+          "Agreement can only be suspended by the consumer or producer."
+        );
+      }
```
This shall not be needed, we already call `assertRequesterIsConsumerOrProducer` above

AsterITA replied:
> This shall not be needed, we already call `assertRequesterIsConsumerOrProducer` above

Right, thanks, I would still need to use one of those check when creating the eventType, I'll change it soon!

ecamellini replied:
True, you could `throw genericError` instead of `new Error`, so that we use the errors we defined in `models` (even though this error will never happen...)

ecamellini commented on packages/agreement-process/src/services/agreementSuspensionProcessor.ts:
```diff
@@ -111,39 +68,15 @@ export async function suspendAgreementLogic({
     ...updateSeed,
   };
-
-  if (authData.organizationId === agreement.data.producerId) {
-    return [
-      updatedAgreement,
-      toCreateEventAgreementSuspendedByProducer(
-        updatedAgreement,
-        agreement.metadata.version,
-        correlationId
-      ),
-    ];
-  } else if (authData.organizationId === agreement.data.consumerId) {
-    return [
-      updatedAgreement,
-      toCreateEventAgreementSuspendedByConsumer(
-        updatedAgreement,
-        agreement.metadata.version,
-        correlationId
-      ),
-    ];
-  } else {
-    throw new Error(
-      "Unexpected organizationId: Agreement can be suspended only by consumer or producer"
-    );
-  }
```
Why not keeping the event creation here as you do in the other processor methods?

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -324,17 +577,56 @@ export function agreementServiceBuilder(
+        );
+      }
+
+      const eventType = isProducer
+        ? toCreateEventAgreementSuspendedByProducer
+        : toCreateEventAgreementSuspendedByConsumer;
+
+      await repository.createEvent(
+        eventType(updatedAgreement, agreement.metadata.version, correlationId)
+      );
```
Left a comment also in the corresponding processor file: I would keep the creation of the event in the processor, as we are doing for other processors e.g., submission, upgrade, etc. and just call the `.createEvent` on the return value here

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -324,17 +577,56 @@ export function agreementServiceBuilder(
-        eserviceQuery,
-        correlationId,
+        descriptor,
+        consumer: await retrieveTenant(agreement.data.consumerId, tenantQuery),
```
Same comment left above: I would create a const and pass it here

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -357,82 +649,176 @@ export function agreementServiceBuilder(
+
+      const [updatedAgreement, updatesEvents] = await processActivateAgreement({
+        agreementData: agreement,
+        eservice: await retrieveEService(
+          agreement.data.eserviceId,
+          eserviceQuery
+        ),
```
Same: create a const and pass it here

ecamellini commented on packages/agreement-process/src/services/agreementActivationProcessor.ts:
```diff
@@ -48,75 +40,43 @@ import {
+}): Promise<[Agreement, Array<CreateEvent<AgreementEvent>>]> {
+  const agreement = agreementData.data;
 
+  const consumer = await retrieveTenant(agreement.consumerId, tenantQuery);
```
Same question I made in the other processor: why are other entities (Eservices, Agreements) retrieved in `agreementService.ts` while Tenants (producer/consumer) are retrieved here in the processor?

AsterITA replied:
> Same question I made in the other processor: why are other entities (Eservices, Agreements) retrieved in `agreementService.ts` while Tenants (producer/consumer) are retrieved here in the processor?

See my comment [here](https://github.com/pagopa/interop-be-monorepo/pull/500#discussion_r1601287412)

ecamellini reviewed (commented):
Left some comments, plus a general consideration: the fact that you had to update some tests means that we are changing the order of assertions in some way, thus the errors we return in some cases - is it ok? Could I ask to reorder things so that tests pass without changes?

AsterITA commented:
> Left some comments, plus a general consideration: the fact that you had to update some tests means that we are changing the order of assertions in some way, thus the errors we return in some cases - is it ok? Could I ask to reorder things so that tests pass without changes?

Thanks! I'll try to address them all soon.

With this refactor I moved the retrieve of the objects we need just before they are actually needed which I find desirable, doing so the order of assertions is changed and I changed the test accordingly.

The errors we are returning are the same as before, the only difference is that some errors are thrown further down the logic then before.

If it facilitates the review I could reorder the retrieve as it was before, but I would need to change it back after the reviews are done

taglioni-r commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -263,39 +446,112 @@ export function agreementServiceBuilder(
+        eserviceQuery
+      );
+
+      const descriptor = eservice.descriptors.find(
+        (d) => d.id === agreementToBeCloned.data.descriptorId
+      );
+      assertDescriptorExist(
+        eservice.id,
+        agreementToBeCloned.data.descriptorId,
+        descriptor
+      );
```
Maybe we could create `retrieveDescriptor`, as we already did in catalog.

taglioni-r commented on packages/agreement-process/src/services/agreementSuspensionProcessor.ts:
```diff
@@ -111,39 +75,37 @@ export async function suspendAgreementLogic({
     );
   }
+
+  const eventType = isProducer
+    ? toCreateEventAgreementSuspendedByProducer
+    : toCreateEventAgreementSuspendedByConsumer;
+
+  return eventType(updatedAgreement, agreement.metadata.version, correlationId);
```
Could we avoid this and make it more verbose, like the following?
```
return isProducer
     ? toCreateEventAgreementSuspendedByProducer(updatedAgreement, agreement.metadata.version, correlationId)
     : toCreateEventAgreementSuspendedByConsumer(updatedAgreement, agreement.metadata.version, correlationId);
```

MalpenZibo reviewed (approved):
:+1:

galales reviewed (approved):
LGTM, just minor considerations.

(❤️  the cleaning using `retrieve*` functions)

ecamellini commented on packages/agreement-process/src/services/agreementService.ts:
```diff
@@ -467,451 +1021,58 @@ async function createAndCopyDocumentsForClonedAgreement(
-    tenantQuery: TenantQuery;
-    eserviceQuery: EserviceQuery;
-  },
+export function createArchivedAgreementEvent(
```
```suggestion
export function createAgreementArchivedByUpgradeEvent(
```

ecamellini reviewed (approved):
If I understood correctly, now all the retrievals, validations, and business logic happen in the main `agreementService.ts` file, while the processor files exist only to create the right update seeds/events. For me, it works also this way - still, I leave my two cents below, not a strong opinion :)

 Personally, I would have preferred an approach where:
- in the `agreementService.ts` we perform all and only the retrievals from the readmodel
- we call the processor/logic methods that contain all validations, all the business logic, and return the events to create
- back in the `agreementService.ts` we create the events in the event store

In this way, the `agreementService.ts` just acts as an orchestrator for the interactions with the readmodel and the event store repository, and the file gets smaller/cleaner while the processors contain business logic and validations.
